### ISSUE TITLE
Transaction error: "WARNING:  there is no transaction in progress"

### ISSUE THREAD

**Issue by nozer** [balderdashy/sails-postgresql#136]
I am performing a transaction via sails waterline on postgres by using something like below and I am getting a "WARNING:  there is no transaction in progress" error in the postgres logs: 

```
SomeModel.checkSomething()
.then(Db.startTransaction)
.return(someData)
.then(SomeModel.create)
...
.then(Db.commit, Db.rollback)
...
```

Here is the code for startTransaction, commit, and rollback in a model named Db.js

```
'use strict';
var Bluebird = require('bluebird');
module.exports = {

   startTransaction: function(arg){
      return Db.run('start transaction').return(arg);
   },
   rollback: function(err){
      return Db.run('rollback').then(function(e){
         //console.log(e, 'rollback err');
         // let's just pass on the reason that caused this rollback, if any.
         err = err || new Error("Database rollback");
         return Promise.reject(err);
      });
   },
   commit: function(arg){
      return Db.run('commit').return(arg);
   },
   run: function(q, params){
      if (params) {
         return Bluebird.promisify(Db.query)(q, params);
      }
      return Bluebird.promisify(Db.query)(q);
   }
};
```

And here are all the logs (here in a prettier format: https://bpaste.net/show/06bb02e11cba) 
I think the problem may be related to different session id  for commit (in comparison to the session id for start transaction). If so, how can I prevent a different session id from being given? If not, why am I getting this there is no transaction in progress error? (log file is shown in its entirety from the moment the db server is started)

```
2015-02-19 13:32:19 UTC 54e5e5e3.1cb9 [7353-1] LOG:  database system was shut down at 2015-02-19 13:32:18 UTC
2015-02-19 13:32:19 UTC 54e5e5e3.1cb7 [7351-3] LOG:  database system is ready to accept connections
2015-02-19 13:32:19 UTC 54e5e5e3.1cbd [7357-1] LOG:  autovacuum launcher started
2015-02-19 13:32:19 UTC 54e5e5e3.1cbf [7359-1] [unknown]@[unknown] LOG:  incomplete startup packet
2015-02-19 13:32:20 UTC 54e5e5e4.1cc2 [7362-1] postgres@postgres LOG:  statement: SELECT d.datname as "Name",
	       pg_catalog.pg_get_userbyid(d.datdba) as "Owner",
	       pg_catalog.pg_encoding_to_char(d.encoding) as "Encoding",
	       d.datcollate as "Collate",
	       d.datctype as "Ctype",
	       pg_catalog.array_to_string(d.datacl, E'\n') AS "Access privileges"
	FROM pg_catalog.pg_database d
	ORDER BY 1;
2015-02-19 13:32:20 UTC 54e5e5e4.1cc5 [7365-1] postgres@postgres LOG:  statement: SELECT d.datname as "Name",
	       pg_catalog.pg_get_userbyid(d.datdba) as "Owner",
	       pg_catalog.pg_encoding_to_char(d.encoding) as "Encoding",
	       d.datcollate as "Collate",
	       d.datctype as "Ctype",
	       pg_catalog.array_to_string(d.datacl, E'\n') AS "Access privileges"
	FROM pg_catalog.pg_database d
	ORDER BY 1;
2015-02-19 13:32:21 UTC 54e5e5e5.1cc8 [7368-1] postgres@postgres LOG:  statement: SELECT d.datname as "Name",
	       pg_catalog.pg_get_userbyid(d.datdba) as "Owner",
	       pg_catalog.pg_encoding_to_char(d.encoding) as "Encoding",
	       d.datcollate as "Collate",
	       d.datctype as "Ctype",
	       pg_catalog.array_to_string(d.datacl, E'\n') AS "Access privileges"
	FROM pg_catalog.pg_database d
	ORDER BY 1;
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-1] vdsr@sails LOG:  execute <unnamed>: SELECT "users"."firstName", "users"."middleName", "users"."lastName", "users"."email", "users"."password", "users"."resetKey", "users"."resetKeyGeneratedAt", "users"."isMainUser", "users"."companyRef", "users"."id", "users"."createdAt", "users"."updatedAt" FROM "users" AS "users"  WHERE "users"."id" = $1  LIMIT 1
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-2] vdsr@sails DETAIL:  parameters: $1 = '8'
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-3] vdsr@sails LOG:  execute <unnamed>: select p.name from permissions p  inner join rolepermissions rp on rp."permissionRef" = p.id   inner join roles r on r.id = rp."roleRef"  inner join userroles ur on ur."roleRef" = r.id  where ur."userRef" = $1
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-4] vdsr@sails DETAIL:  parameters: $1 = '8'
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-5] vdsr@sails LOG:  statement: start transaction
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-6] vdsr@sails LOG:  execute <unnamed>: INSERT INTO "groups" ("name", "notes", "parentGroupRef", "companyRef", "createdAt", "updatedAt") values ($1, $2, $3, $4, $5, $6) RETURNING *
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-7] vdsr@sails DETAIL:  parameters: $1 = 'dsadsa', $2 = 'dsadasd', $3 = '1', $4 = '1', $5 = '2015-02-19 13:32:37+00', $6 = '2015-02-19 13:32:37+00'
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-8] vdsr@sails LOG:  execute <unnamed>: SELECT "groups"."name", "groups"."notes", "groups"."parentGroupRef", "groups"."isCompany", "groups"."hierPos", "groups"."companyRef", "groups"."id", "groups"."createdAt", "groups"."updatedAt" FROM "groups" AS "groups"  WHERE "groups"."companyRef" = $1 
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-9] vdsr@sails DETAIL:  parameters: $1 = '1'
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-10] vdsr@sails LOG:  statement:  update groups set "hierPos" = case id when 2 then '1' when 3 then '2' when 4 then '3' when 5 then '4' when 6 then '4.1' when 10 then '4.1.1' when 18 then '4.2' when 13 then '5' when 17 then '6' when 19 then '7' when 20 then '8' when 21 then '9' else "hierPos" end  where id in (2,3,4,5,6,10,18,13,17,19,20,21) 
2015-02-19 13:32:37 UTC 54e5e5f5.1cda [7386-11] vdsr@sails LOG:  statement: commit
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-1] vdsr@sails LOG:  execute <unnamed>: SELECT "users"."firstName", "users"."middleName", "users"."lastName", "users"."email", "users"."password", "users"."resetKey", "users"."resetKeyGeneratedAt", "users"."isMainUser", "users"."companyRef", "users"."id", "users"."createdAt", "users"."updatedAt" FROM "users" AS "users"  WHERE "users"."id" = $1  LIMIT 1
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-2] vdsr@sails DETAIL:  parameters: $1 = '8'
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-3] vdsr@sails LOG:  execute <unnamed>: select p.name from permissions p  inner join rolepermissions rp on rp."permissionRef" = p.id   inner join roles r on r.id = rp."roleRef"  inner join userroles ur on ur."roleRef" = r.id  where ur."userRef" = $1
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-4] vdsr@sails DETAIL:  parameters: $1 = '8'
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-5] vdsr@sails LOG:  statement: start transaction
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-6] vdsr@sails LOG:  execute <unnamed>: INSERT INTO "groups" ("name", "notes", "isCompany", "parentGroupRef", "companyRef", "createdAt", "updatedAt") values ($1, $2, $3, $4, $5, $6, $7) RETURNING *
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-7] vdsr@sails DETAIL:  parameters: $1 = 'fdsfds', $2 = 'fdsfsd', $3 = 't', $4 = '1', $5 = '1', $6 = '2015-02-19 13:33:21+00', $7 = '2015-02-19 13:33:21+00'
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-8] vdsr@sails LOG:  execute <unnamed>: INSERT INTO "users" ("firstName", "lastName", "email", "companyRef", "isMainUser", "password", "middleName", "createdAt", "updatedAt") values ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING *
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-9] vdsr@sails DETAIL:  parameters: $1 = 'fds', $2 = 'fds', $3 = 'fdsf@ds.com', $4 = '22', $5 = 't', $6 = 'bcaebfa08ec1ecfc6446c9bee391aa68', $7 = '', $8 = '2015-02-19 13:33:21+00', $9 = '2015-02-19 13:33:21+00'
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-10] vdsr@sails LOG:  execute <unnamed>: INSERT INTO "roles" ("name", "companyRef", "createdAt", "updatedAt") values ($1, $2, $3, $4) RETURNING *
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-11] vdsr@sails DETAIL:  parameters: $1 = 'Master', $2 = '22', $3 = '2015-02-19 13:33:21+00', $4 = '2015-02-19 13:33:21+00'
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-12] vdsr@sails LOG:  execute <unnamed>: INSERT INTO "userroles" ("roleRef", "userRef", "createdAt", "updatedAt") values ($1, $2, $3, $4) RETURNING *
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-13] vdsr@sails DETAIL:  parameters: $1 = '14', $2 = '18', $3 = '2015-02-19 13:33:21+00', $4 = '2015-02-19 13:33:21+00'
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-14] vdsr@sails LOG:  execute <unnamed>: INSERT INTO "rolepermissions" ("roleRef", "permissionRef", "createdAt", "updatedAt") values ($1, $2, $3, $4) RETURNING *
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-15] vdsr@sails DETAIL:  parameters: $1 = '14', $2 = '24', $3 = '2015-02-19 13:33:21+00', $4 = '2015-02-19 13:33:21+00'
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-16] vdsr@sails LOG:  execute <unnamed>: INSERT INTO "rolepermissions" ("roleRef", "permissionRef", "createdAt", "updatedAt") values ($1, $2, $3, $4) RETURNING *
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-17] vdsr@sails DETAIL:  parameters: $1 = '14', $2 = '22', $3 = '2015-02-19 13:33:21+00', $4 = '2015-02-19 13:33:21+00'
2015-02-19 13:33:21 UTC 54e5e621.1ce3 [7395-1] vdsr@sails LOG:  execute <unnamed>: SELECT "groups"."name", "groups"."notes", "groups"."parentGroupRef", "groups"."isCompany", "groups"."hierPos", "groups"."companyRef", "groups"."id", "groups"."createdAt", "groups"."updatedAt" FROM "groups" AS "groups"  WHERE "groups"."companyRef" = $1 
2015-02-19 13:33:21 UTC 54e5e621.1ce3 [7395-2] vdsr@sails DETAIL:  parameters: $1 = '1'
2015-02-19 13:33:21 UTC 54e5e621.1ce2 [7394-18] vdsr@sails LOG:  statement:  update groups set "hierPos" = case id when 2 then '1' when 3 then '2' when 4 then '3' when 5 then '4' when 6 then '4.1' when 10 then '4.1.1' when 18 then '4.2' when 13 then '5' when 17 then '6' when 19 then '7' when 20 then '8' when 21 then '9' else "hierPos" end  where id in (2,3,4,5,6,10,18,13,17,19,20,21) 
2015-02-19 13:33:21 UTC 54e5e621.1ce3 [7395-3] vdsr@sails LOG:  statement: commit
2015-02-19 13:33:21 UTC 54e5e621.1ce3 [7395-4] vdsr@sails WARNING:  there is no transaction in progress
```

UPDATE: 
Cause of this seems to be a new session is being created by the time commit query is issued. Based on the code in the sails postgresql, it seems that a new connection is being spawned in each call to model methods, hence the change in session. Under these conditions, it seems impossible to do transactions in sails.

**Comment by tjwebb** [comment]
> 
```
SomeModel.checkSomething()
.then(Db.startTransaction)
.return(someData)
.then(SomeModel.create)
...
.then(Db.commit, Db.rollback)
```

I don't see how this could work. Every time you make a db query from node, you aren't even guaranteed that it'll use the same database connection -- it's going to pick one from the connection pool. I'm having trouble seeing how you would achieve transactionality with this approach.

**Comment by tjwebb** [comment]
> Under these conditions, it seems impossible to do transactions in sails.

Using Promises, you couldn't achieve transactionality at all, regardless of framework. It's not possible to begin a transaction, send a query to the database, and then serialize data out of the database in the middle of a transaction. You could then alter that data during the rest of the transaction, and then the data you have in node-land would be stale.  Asynchronicity is fundamentally at odds with a transactional database. 

The only way to do what you are trying to do is send a giant SQL blob to be run as a single query.

**Comment by nozer** [comment]
Yes, and that would mean that I would have to generate all my queries manually (without using the model methods) and send them as a giant SQL blob as you advise. At that point, model methods such as .create, .update are not useful. 

Is there any model method like .getNativeSqlQuery() that we can issue on model methods like 
Users.create(dataObj).getNativeSqlQuery() that would return the underlying query that would be executed?

**Comment by tjwebb** [comment]
I'm not sure if there's a public method in Waterline to expose the sql. @particlebanana any quick pointers here?

**Comment by jub3i** [comment]
I think the method your looking for is http://sailsjs.org/#!/documentation/reference/waterline/models/query.html 

I started this https://github.com/balderdashy/sails-mysql/wiki/Transactions to provide for transactions in mySQL but its pretty rough, and could definitely use some improvement.

**Comment by moglash** [comment]
:+1: 
I was having really bad time because of the unfinished transactions.
+1 Exposing the native sql query would be really nice.
Another possible fix for the problem is using an optional connection parameter in the query. 
For eg. :
```javascript

var pg = require('pg');
var conString = "postgres://username:password@localhost/database";

var client = new pg.Client(conString);
client.connect(function(err) {
  if(err) {
    return console.error('could not connect to postgres', err);
  }
  Model.create(data, client, cb);
  ....
  Model.query(sql, client, cb);
  client.end();
 
});
```

**Comment by moglash** [comment]
@nozer i've made a simple promise based transaction flow example: 
I've tested it, and it's working nice for me.
https://gist.github.com/moglash/a478912be460a8790a62

**Comment by fundon** [comment]
@barczaG 404, could you post a new gist?

**Comment by barczaG** [comment]
@fundon I've updated the link. If you interested I can rewrite it using `co` (generator based workflow)

**Comment by fundon** [comment]
@barczaG thx
I'm interesting. I'm looking forward that the `co` version.

**Comment by sailsbot** [comment]
Thanks for posting, @nozer.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by kevinburkeshyp** [comment]
FWIW I recently open sourced the library we use for doing database transactions in Postgres - same interface as `Model.query`, but they'll all be on the same connection. https://github.com/Shyp/pg-transactions
